src/filters/ExtendedKalmanFilter.h

```diff
@@ -0,0 +1,284 @@
+#pragma once
+
+#include <Eigen/Core>
+#include <Eigen/LU>
+
+#include "KalmanFilterBase.h"
+#include "StateSpaceUtil.h"
+
+/**
+ * Represents a square noise covariance matrix.
+ *
+ * @tparam numStates The number of states in this system.
+ * @tparam size The size of this matrix.
+ * @tparam paramSize The dimension of the vector accepted by get() in addition to a state
+ * vector.
+ */
+template <int numStates, int size, int paramSize> class NoiseCovMat
+{
+public:
+	using state_t = Eigen::Matrix<double, numStates, 1>;
+	using param_t = Eigen::Matrix<double, paramSize, 1>;
+
+	/**
+	 * Create a time-invariant noise covariance matrix modelling independent noise with the
+	 * given standard deviations.
+	 *
+	 * @param stdDevs The standard deviations of each element.
+	 */
+	explicit NoiseCovMat(const Eigen::Matrix<double, size, 1> &stdDevs)
+		: NoiseCovMat(StateSpace::createCovarianceMatrix(stdDevs))
+	{
+	}
+
+	/**
+	 * Create a time-invariant noise covariance matrix equal to the given matrix.
+	 *
+	 * @param mat The noise covariance matrix.
+	 */
+	NoiseCovMat(const Eigen::Matrix<double, size, size> &mat)
+		: func([mat](const state_t &x, const param_t &param) { return mat; })
+	{
+	}
+
+	/**
+	 * Create a time-varying noise covariance matrix. At runtime, the matrix will be calculated
+	 * when needed using the supplied function.
+	 *
+	 * @param func The function that supplies the noise covariance matrix,
+	 * given the state vector and one additional vector. For process noise, this is the input
+	 * vector. For output noise, this is the output vector.
+	 */
+	NoiseCovMat(const std::function<Eigen::Matrix<double, size, size>(const state_t &,
+																	  const param_t &)> &func)
+		: func(func)
+	{
+	}
+
+	/**
+	 * Gets the noise covariance matrix, given the current state and additonal parameter.
+	 * The matrix may be time-invariant, which case the values of x and param do not matter.
+	 *
+	 * @param x The current state vector.
+	 * @param param The parameter vector, as defined by the use of this matrix.
+	 * @return The noise covariance matrix.
+	 */
+	Eigen::Matrix<double, size, size> get(const state_t &x, const param_t &param)
+	{
+		return func(x, param);
+	}
+
+private:
+	std::function<Eigen::Matrix<double, size, size>(const state_t &, const param_t &)> func;
+};
+
+/**
+ * Implements a discrete-time EKF. This implements the more general system form described here:
+ * https://en.wikipedia.org/wiki/Extended_Kalman_filter#Non-additive_noise_formulation_and_equations
+ *
+ * @tparam numStates The number of states in the system.
+ * @tparam numInputs The number of inputs in the system.
+ * @tparam numOutputs The number of outputs in the system.
```

Comment:
It might be more accurate to call this e.g. the dimension of the state space of the system. (Technically there are an infinite number of states.)